File: Arrays/1_Two_Sums/TwoSums.py
# Method 1 : Using 2 forloops, take the sum of an element i with each of the element j and compare the sum to the target value to determine the index of the elements that add to it.
# Time complexity : O(n^2)
def two_sums(nums : list[int], target : int):
    for i in range(len(nums)):
        for j in range(i+1,len(nums)):
           if target == (nums[i]+nums[j]):
               return i, j

# Method 2 :  Use hashing table - take the difference between "each element and the target value", check if the difference exist in the dictionary (a.k.a : hashtable).
#                                 if it doesn't exist, add the element and its index into the hashtable.
# Time Complexity : O(n)
def hashing_two_sums(nums, target):
    h = dict()
    for i, num in enumerate(nums):
        if target - num in h:
            return [h[target - num], i]
        h[num] = i



if __name__ == "__main__":
    nums = [2,7,11,15]
    target = 9
    print(f"[Method 1] : index of the sum of the target({target}) = {two_sums(nums = nums, target = target)}")
    print(f"[Method 2] : index of the sum of the target({target}) = {hashing_two_sums(nums, target)} ")

File: Arrays/Easy/1_Two_Sums/TwoSums.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution{
    public :
        // Method 1: Uses 2 forloops and in a forwarding method, we take the difference of each element i to the target value and check if the difference results in it being the element j.
        vector <int> TwoSums(vector<int> &nums, int target){
            vector<int> sum;

            // Idea : take target value - nums[i] for which the reuslt will equal to one of the nums[j] 
            // Signifying that nums[j] is the value
            for(int i = 0; i < nums.size(); i++){
                for(int j = i+1; j < nums.size(); j++){
                    if(nums[j] == (target - nums[i])){
                        sum.push_back(i);
                        sum.push_back(j);
                        return sum;
                    }
                }
            }
            return sum;
        }
        // Method 2 : Take the difference between each element from the number array and the target value and check if the difference exist as element in the hash table.
        //            if not, add the current element into the hash table.
        vector <int> Hashmap_TwoSums(vector<int> &nums, int target){
            unordered_map<int,int> m;
            vector<int> temp;
            for(int i = 0; i < nums.size(); ++i){
                int diff = target - nums[i];
                unordered_map<int,int>::iterator it = m.find(diff);
                if(it != m.end()){
                    temp.push_back(it->second);
                    temp.push_back(i);
                    return temp;
                }
                m[nums[i]] = i;
            }
            return temp;
        }

};

void display(const vector<int> &vec){
    vector<int>::const_iterator it = vec.begin();
    while(it != vec.end()){
        cout << *it << " ";
        it++;
    }
    cout << endl;
}

int main(){
    vector<int> x;
    x.push_back(2);
    x.push_back(7);
    x.push_back(11);
    x.push_back(15);
    x.push_back(4);
    x.push_back(5);

    int target = 9;

    Solution S;

    // Method 1
    vector<int> temp = S.TwoSums(x, target);
    display(temp);

    // Method 2
    vector<int> temp_2 = S.Hashmap_TwoSums(x, target);
    display(temp_2);





    return 0;
}